### DISCUSSION TITLE
How to use functions like `np.unique()` and `np.any()` on `pybind11::array`?

### DISCUSSION THREAD

**Discussion by sun1638650145** [pybind/pybind11#3184] [General]
I want to use functions similar to `np.unique()` and `np.any()` on `pybind11::array`, so I first cast `pybind11::array` to `Eigen::ArrayXf`, Then proceed with processing. But no matter how I operate, I can only process `1/2-D array`, but I also expect to be able to process `3D array`, is it my idea is wrong, or `pybind11` provides built-in functions.

```c++
void MyUnique(const pybind11::array &array) {
    auto arr = pybind11::cast<Eigen::ArrayXf>(array);

    // Then use std::unique and std::sort to remove duplicates.
}

```
Below is the function I expect.

```python
unique_ = np.unique(y)
any_ = np.any(y != y.astype(int))
```

English is not my native language; please excuse typing errors.

**Comment by Skylion007** [comment]
You can always just import numpy through it's python interface to access it's advanced functionality. 
``` 
auto np = py::module_::import("numpy");
auto out = np.attr("unique")(array);
```
See the documentation on importing Python modules for mode details: https://pybind11.readthedocs.io/en/stable/advanced/embedding.html?highlight=import#importing-modules

Alternatively, Eigen has it's own unique and sort functions etc so I would just use those on the array object.

**Reply by sun1638650145** [reply]
First, thank you very much, but it seems that I did not express what I mean clearly. I expect is to use pybind11::array to achieve similar functions, not directly from numpy.

**Reply by Skylion007** [reply]
We don't have those functions exposed via the C++ API through py:array.

**Reply by sun1638650145** [reply]
Because I just need to remove the duplicated values ​​of the pybind11::array, and don’t need the original shape, can we cast the array to a flat std::vector?

**Reply by Skylion007** [reply]
Yes, there is a caster from array to std::vector. It also has STL compatible iterators so you can call std::unique etc...

**Reply by sun1638650145** [reply]
I tried to convert to std::vector but 2-D and 3-D does not seem to be able to use pybind11::cast. What should I do?
